lib/repl.js

```diff
@@ -585,10 +595,21 @@ REPLServer.prototype.createContext = function() {
     context = global;
   } else {
     context = vm.createContext();
-    for (var i in global) context[i] = global[i];
-    context.console = new Console(this.outputStream);
     context.global = context;
     context.global.global = context;
+    const _console = new Console(this.outputStream);
+    Object.defineProperty(context, 'console', {
+      configurable: true,
+      enumerable: true,
+      get: () => _console
+    });
+    Object.getOwnPropertyNames(global).filter((name) => {
+      if (name === 'console' || name === 'global') return false;
+      return GLOBAL_OBJECT_PROPERTY_MAP[name] === undefined;
```

Comment:
Yes, `GLOBAL_OBJECT_PROPERTIES` is just a map used for convenience and to avoid having to scan the array each time for a lookup on `name`. So, `GLOBAL_OBJECT_PROPERTY_MAP[name] === undefined` is essentially equivalent to `GLOBAL_OBJECT_PROPERTIES.indexOf(name) === -1`.

However, I'm not seeing the false positives from `Object.prototype` that you mention. A tab completion does exist for `constructor`, but that is the case for `node` version 6.2.2 as well.

    ~ ❯❯❯ which node
    /Users/lanceball/.nvm/versions/node/v6.2.2/bin/node
    ~ ❯❯❯ node
    > con
    const        continue
    
    console
    
    constructor
    
    > con

And `constructor` doesn't appear in `global`'s own property names.

    > Object.getOwnPropertyNames(global).filter((name) => name === 'constructor');
    [] 

Do I misunderstand your comment?